cli/flox-rust-sdk/src/providers/services.rs

```diff
@@ -1,8 +1,207 @@
+// write it out to either the env var location or a temp file
+pub fn write_process_compose_config(
+    config: &ProcessComposeConfig,
+    path: impl AsRef<Path>,
+) -> Result<(), ServiceError> {
+    let contents = serde_yml::to_string(&config).map_err(ServiceError::GenerateConfig)?;
+    std::fs::write(path, contents).map_err(ServiceError::WriteConfig)?;
+    Ok(())
+}
+
+/// A container for the path to the `process-compose` config file
+///
+/// This is necessary in the situation where the config file is written to a temporary location
+/// and we need to keep the destructor from deleting the temp directory before we're done with it.
+#[derive(Debug)]
+pub struct ConfigPath {
+    pub path: PathBuf,
+    _tempdir: Option<tempfile::TempDir>,
+}
+
+/// Determines the location to write the `process-compose` config file
+pub fn process_compose_config_write_location() -> Result<ConfigPath, ServiceError> {
+    if let Ok(path) = env::var(SERVICES_TEMP_CONFIG_PATH_VAR) {
+        return Ok(ConfigPath {
+            path: PathBuf::from(path),
+            _tempdir: None,
+        });
+    }
+    let temp_dir = tempfile::tempdir().map_err(ServiceError::WriteConfig)?;
+    let path = temp_dir.path().join("process-compose.yaml");
+    Ok(ConfigPath {
+        path,
+        _tempdir: Some(temp_dir),
+    })
+}
+
+pub fn maybe_make_service_config_file(
+    flox: &Flox,
+    lockfile: &LockedManifestCatalog,
+) -> Result<Option<ConfigPath>, ServiceError> {
+    let service_config_path = if flox.features.services {
+        let config_path = process_compose_config_write_location()?;
+        write_process_compose_config(
+            &lockfile.manifest.services.clone().into(),
+            &config_path.path,
+        )?;
+        tracing::debug!(
+            path = traceable_path(&config_path.path),
+            "wrote service config"
+        );
+        Some(config_path)
+    } else {
+        None
+    };
+    Ok(service_config_path)
+}
+
+#[cfg(test)]
+mod tests {
+    use std::str::FromStr;
+
+    use indoc::indoc;
+    use proptest::prelude::*;
+
+    use super::*;
+    use crate::flox::test_helpers::flox_instance_with_optional_floxhub_and_client;
+    use crate::flox::{EnvironmentName, EnvironmentOwner};
+    use crate::models::environment::path_environment::{InitCustomization, PathEnvironment};
+    use crate::models::environment::{Environment, PathPointer};
+
+    proptest! {
+        #[test]
+        fn test_process_compose_config_roundtrip(config: ProcessComposeConfig) {
+            let path = process_compose_config_write_location().unwrap();
+            write_process_compose_config(&config, &path.path).unwrap();
+            let contents = std::fs::read_to_string(&path.path).unwrap();
+            let deserialized: ProcessComposeConfig = serde_yml::from_str(&contents).unwrap();
+            prop_assert_eq!(config, deserialized);
+        }
+    }
+
+    #[test]
+    fn built_environments_generate_service_config() {
```

Comment:
nonblocking question: this seems to be testing the logic on `LockedEnvironment::build` as the caller of logic in this file - would it make more sense for this test to live in the same file as `build`? I don't have a strong opinion

Reply:
Also a bit indifferent, but probably. I assume you meant `CoreEnvironment::build` rather than `LockedManifest::build`? I've moved it in 0cd21bbb.